checker/src/org/checkerframework/checker/index/searchindex/SearchIndexAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,152 @@
+ * methods. It is part of the Index Checker.
+ */
+public class SearchIndexAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
+
+    public final AnnotationMirror UNKNOWN, BOTTOM;
+
+    public SearchIndexAnnotatedTypeFactory(BaseTypeChecker checker) {
+        super(checker);
+        UNKNOWN = AnnotationUtils.fromClass(elements, SearchIndexUnknown.class);
+        BOTTOM = AnnotationUtils.fromClass(elements, SearchIndexBottom.class);
+        this.postInit();
+    }
+
+    /**
+     * Provides a way to query the Constant Value Checker, which computes the values of expressions
+     * known at compile time (constant propagation and folding).
+     */
+    ValueAnnotatedTypeFactory getValueAnnotatedTypeFactory() {
+        return getTypeFactoryOfSubchecker(ValueChecker.class);
+    }
+
+    @Override
+    protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
+        return new LinkedHashSet<>(
+                Arrays.asList(
+                        SearchIndexFor.class,
+                        SearchIndexBottom.class,
+                        SearchIndexUnknown.class,
+                        NegativeIndexFor.class));
+    }
+
+    @Override
+    public QualifierHierarchy createQualifierHierarchy(
+            MultiGraphQualifierHierarchy.MultiGraphFactory factory) {
+        return new SearchIndexQualifierHierarchy(factory);
+    }
+
+    private final class SearchIndexQualifierHierarchy extends MultiGraphQualifierHierarchy {
+
+        public SearchIndexQualifierHierarchy(
+                MultiGraphQualifierHierarchy.MultiGraphFactory factory) {
+            super(factory);
+        }
+
+        @Override
+        public AnnotationMirror greatestLowerBound(AnnotationMirror a1, AnnotationMirror a2) {
+            if (AnnotationUtils.areSame(a1, UNKNOWN)) {
+                return a2;
+            }
+            if (AnnotationUtils.areSame(a2, UNKNOWN)) {
+                return a1;
+            }
+            if (AnnotationUtils.areSame(a1, BOTTOM)) {
+                return a1;
+            }
+            if (AnnotationUtils.areSame(a2, BOTTOM)) {
+                return a2;
+            }
+            if (isSubtype(a1, a2)) {
+                return a1;
+            }
+            if (isSubtype(a2, a1)) {
+                return a2;
+            }
+            return BOTTOM;
+        }
+
+        @Override
+        public AnnotationMirror leastUpperBound(AnnotationMirror a1, AnnotationMirror a2) {
+            if (AnnotationUtils.areSame(a1, UNKNOWN)) {
+                return a1;
+            }
+            if (AnnotationUtils.areSame(a2, UNKNOWN)) {
+                return a2;
+            }
+            if (AnnotationUtils.areSame(a1, BOTTOM)) {
+                return a2;
+            }
+            if (AnnotationUtils.areSame(a2, BOTTOM)) {
+                return a1;
+            }
+            if (isSubtype(a1, a2)) {
```

Comment:
This isn't right. The LUB of @SearchIndex({"a", "b"}) and @NegativeIndexFor("a") is @SearchIndex("a").